shell/platform/windows/windows_lifecycle_manager.cc

```diff
@@ -0,0 +1,84 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "windows_lifecycle_manager.h"
+
+#include <TlHelp32.h>
+#include <WinUser.h>
+#include <Windows.h>
+#include <tchar.h>
+
+#include "flutter/shell/platform/windows/flutter_windows_engine.h"
+
+namespace flutter {
+
+WindowsLifecycleManager::WindowsLifecycleManager(FlutterWindowsEngine& engine)
+    : engine_(engine) {}
+
+WindowsLifecycleManager::~WindowsLifecycleManager() {}
+
+void WindowsLifecycleManager::Quit(int64_t exit_code) {
+  PostQuitMessage(exit_code);
+}
+
+bool WindowsLifecycleManager::WindowProc(HWND hwnd,
+                                         UINT msg,
+                                         WPARAM wpar,
+                                         LPARAM lpar,
+                                         LRESULT* result) {
+  switch (msg) {
+    case WM_CLOSE:
+      if (IsLastWindowOfProcess()) {
+        engine_.RequestApplicationQuit(PlatformHandler::kExitTypeCancelable,
+                                       wpar);
+      }
+      return true;
+  }
+  return false;
+}
+
+static BOOL CALLBACK WindowEnumCallback(HWND hwnd, LPARAM user_data) {
+  HWND parent = GetParent(hwnd);
+  if (parent == NULL) {
+    int64_t& count = *static_cast<int64_t*>(reinterpret_cast<void*>(user_data));
+    count++;
+  }
+  return true;
+}
+
+bool WindowsLifecycleManager::IsLastWindowOfProcess() {
+  DWORD pid = GetCurrentProcessId();
+  HANDLE thread_snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPTHREAD, 0);
+  if (thread_snapshot == INVALID_HANDLE_VALUE) {
+    FML_LOG(ERROR) << "Failed to get threads snapshot";
+    return true;
+  }
+
+  THREADENTRY32 thread;
+  thread.dwSize = sizeof(thread);
+  if (!Thread32First(thread_snapshot, &thread)) {
+    DWORD error_num = GetLastError();
+    char msg[256];
+    FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+                   NULL, error_num, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+                   msg, 256, nullptr);
+    FML_LOG(ERROR) << "Failed to get thread(" << error_num << "): " << msg;
+    CloseHandle(thread_snapshot);
+    return true;
+  }
```

Comment:
I'd be tempted to extract this code to a private method like `GetFirstThreadForProcess` for readability. Could probably be static since I don't think you're using any fields of this struct here.

Reply:
Ah I see looking at this, you need to keep `thread_snapshot` alive while you poke through the windows list for the thread.

You could probably extract a tiny class like `ThreadSnapshot` with a method like `std::optional<THREADENTRY32> GetFirstThread()` that returns the first thread and have `ThreadSnapshot` close the snapshot handle in its destructor, which has the nice advantage of ensuring that that happens on return without having to manually clean up before each path out of this function.

Reply:
I don't think I can put all of this into a separate function, as we still need to `CloseHandle(thread_snapshot)` after enumerating all threads, but getting the first thread can be extracted.

Reply:
See my note on doing this in the destructor, which gives a nice RAII pattern and avoids the need for the close call in user code altogether.